wp_api/src/auth.rs

```diff
@@ -0,0 +1,132 @@
+use http::HeaderValue;
+use std::sync::{Arc, Mutex};
+
+#[derive(Debug, Clone, uniffi::Enum)]
+pub enum WpAuthentication {
+    AuthorizationHeader { token: String },
+    Bearer { token: String },
+    None,
+}
+
+#[uniffi::export]
+fn wp_authentication_from_username_and_password(
+    username: String,
+    password: String,
+) -> WpAuthentication {
+    WpAuthentication::from_username_and_password(username, password)
+}
+
+impl WpAuthentication {
+    pub fn from_username_and_password(username: String, password: String) -> Self {
+        use base64::prelude::*;
+        WpAuthentication::AuthorizationHeader {
+            token: BASE64_STANDARD.encode(format!("{}:{}", username, password)),
+        }
+    }
+
+    pub fn header_value(&self) -> Option<HeaderValue> {
+        match self {
+            Self::None => None,
+            Self::AuthorizationHeader { token } => {
+                Some(HeaderValue::from_str(&format!("Basic {}", token))
+                .expect("It shouldn't be possible to build WpAuthentication::AuthorizationHeader with an invalid token"))
+            }
+            Self::Bearer { token } => {
+                Some(HeaderValue::from_str(&format!("Bearer {}", token)).expect("It shouldn't be possible to build WpAuthentication::Bearer with an invalid token"))
+            }
+        }
+    }
+}
+
+#[derive(uniffi::Object)]
+pub struct ModifiableAuthenticationProvider {
+    auth: Mutex<WpAuthentication>,
+}
+
+#[uniffi::export]
+impl ModifiableAuthenticationProvider {
+    #[uniffi::constructor]
+    pub fn new(authentication: WpAuthentication) -> Self {
+        Self {
+            auth: Mutex::new(authentication),
+        }
+    }
+
+    pub fn set_authentication(&self, new_authentication: WpAuthentication) {
+        *self.auth.lock().unwrap() = new_authentication;
```

Comment:
Use `expect` instead of `unwrap`?

Reply:
Right.. I looked into this a little bit last week and it looks like unwrapping a lock is the standard practice since in most cases there isn't a clear way to resolve a poison. However, we should use `expect` here as you suggest and put this explanation into the expect message 🤔